lib/wibox/template.lua

```diff
@@ -0,0 +1,439 @@
+-- `wibox.template` :
+--
+-- The module definition should include a default widget and a builder function
+-- that can build the widget with either, the user values or the default.
+--
+--@DOC_wibox_widget_template_concrete_implementation_module_EXAMPLE@
+--
+-- On the user side, it only requires to call the builder function to get an
+-- instance of the widget.
+--
+--@DOC_wibox_widget_template_concrete_implementation_user_EXAMPLE@
+--
+-- @author Aire-One
+-- @copyright 2021 Aire-One <aireone@aireone.xyz>
+--
+-- @widgetmod wibox.template
+-- @supermodule wibox.widget.base
+---------------------------------------------------------------------------
+
+local wbase = require("wibox.widget.base")
+local gtable = require("gears.table")
+local gtimer = require("gears.timer")
+
+local template = {
+    mt = {},
+    queued_updates = {},
+}
+
+local function lazy_load_child(self)
+    if self._private.widget then return self._private.widget end
+
+    local widget_instance = wbase.make_widget_from_value(self._private.widget_template)
+
+    if widget_instance then
+        wbase.check_widget(widget_instance)
+    end
+
+    self._private.widget = widget_instance
+
+    local rem = {}
+
+    for k, conn in ipairs(self._private.connections) do
+        conn.apply()
+
+        if conn.once then
+            if conn.src_obj then
+                conn.src_obj:disconnect_signal("property::"..conn.src_prop, conn.apply)
+            end
+            table.insert(rem, k)
+        end
+    end
+
+    for i = #rem, 1, -1 do
+        table.remove(self._private.connections, rem[i])
+    end
+
+    return self._private.widget
+end
+
+-- Layout this layout.
+-- @method layout
+-- @hidden
+function template:layout(_, width, height)
+    local w = lazy_load_child(self)
+
+    if not w then return end
+
+    return { wbase.place_widget_at(w, 0, 0, width, height) }
+end
+
+-- Fit this layout into the given area.
+-- @method fit
+-- @hidden
+function template:fit(context, width, height)
+    local w = lazy_load_child(self)
+
+    if not w then
+        return 0, 0
+    end
+
+    return wbase.fit_widget(self, context, w, width, height)
+end
+
+function template:draw() end
+
+-- Call the update widget method now and clean the queue for this widget
+-- instance.
+-- @method _do_update_now
+-- @hidden
+function template:_do_update_now()
+    if type(self._private.update_callback) == "function" then
+        self._private.update_callback(self, self._private.update_args)
+    end
+
+    self._private.update_args = nil
+    template.queued_updates[self] = nil
+end
+
+--- Update the widget.
+--
+-- This function will call the `update_callback` function at the end of the
+-- current GLib event loop. Updates are batched by event loop, it means that the
+-- widget can only be update once by event loop. If the `template:update` method
+-- is called multiple times during the same GLib event loop, only the first call
+-- will be run.
```

Comment:
The "end of the event loop" is different from "end of the event loop iteration". The former will only happen when `awesome.quit()` is called and the main loop is shut down.
And "update are batched" is the opposite of "only the first call will be run" when documented that way.

```suggestion
-- Updates are batched. Multiple calls to @{update} within the same event loop
-- iteration will be collected. At the end of the loop iteration, they will be combined into 
-- a single call to the function provided as @{update_callback}.
```